### PR TITLE
Source symlink fix

### LINKED ISSUE #4322: FileExistsError during second cylc install with --directory option
**Describe the bug**

I get traceback when doing a second `cylc install` due to a problem with how the `_cylc-install/source` symlink is handled.

**Release version(s) and/or repository branch(es) affected?**

master

**Steps to reproduce the bug**

1. Install a workflow, e.g. the cylc-doc forecasting workflow
    ```console
    $ cd ~/github/cylc-doc
    $ cylc install -C cylc/doc/etc/tutorial/cylc-forecasting-workflow/
    INSTALLED cylc-forecasting-workflow/run1 from cylc/doc/etc/tutorial/cylc-forecasting-workflow
    ```

1. (Optional - run the workflow, which is what I did but probably isn't needed to reproduce)

1. Install the second run, and get `FileExistsError`
    ```console
    $ cylc install -C cylc/doc/etc/tutorial/cylc-forecasting-workflow/
    Traceback (most recent call last):
    File "~/miniconda3/envs/cylc8/bin/cylc", line 33, in <module>
        sys.exit(load_entry_point('cylc-flow', 'console_scripts', 'cylc')())
    File "~/github/cylc-flow/cylc/flow/scripts/cylc.py", line 538, in main
        execute_cmd(command, *cmd_args)
    File "~/github/cylc-flow/cylc/flow/scripts/cylc.py", line 174, in execute_cmd
        COMMANDS[cmd].resolve()(*args)
    File "~/github/cylc-flow/cylc/flow/terminal.py", line 260, in wrapper
        wrapped_function(*wrapped_args, **wrapped_kwargs)
    File "~/github/cylc-flow/cylc/flow/scripts/install.py", line 187, in main
        install(parser, opts, reg)
    File "~/github/cylc-flow/cylc/flow/scripts/install.py", line 232, in install
        cli_symlink_dirs=cli_symdirs
    File "~/github/cylc-flow/cylc/flow/workflow_files.py", line 1354, in install_workflow
        source_link.symlink_to(source)
    File "~/miniconda3/envs/cylc8/lib/python3.7/pathlib.py", line 1309, in symlink_to
        self._accessor.symlink(target, self, target_is_directory)
    File "~/miniconda3/envs/cylc8/lib/python3.7/pathlib.py", line 418, in symlink
        return os.symlink(a, b)
    FileExistsError: [Errno 17] File exists: 'cylc/doc/etc/tutorial/cylc-forecasting-workflow' -> '~/cylc-run/cylc-forecasting-workflow/_cylc-install/source'
    ```

**Additional context**

The `source` symlink is broken (and was already broken after the first install)
```console
$ readlink ~/cylc-run/cylc-forecasting-workflow/_cylc-install/source
cylc/doc/etc/tutorial/cylc-forecasting-workflow
```

Looks like the relative path needs to be expanded to an absolute one, but this might be a slightly separate problem to the main issue?

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

@datamel I'm happy to have a go at fixing this myself, but can I ask you to try and reproduce?

### PR DESCRIPTION
**Pull request by datamel**
These changes close #4322
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and
  `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

File: cylc/flow/workflow_files.py [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ def install_workflow(
     cylc_install.mkdir(parents=True, exist_ok=True)
     if not source_link.exists():
         install_log.info(f"Creating symlink from {source_link}")
-        source_link.symlink_to(source)
+        source_link.symlink_to(source.resolve())
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:1356]: Tested and this fixes the bug 👍

However, I'm wondering about the `if not source_link.exists()` line - if the source workflow is moved or deleted, then the same issue will resurface (as `broken_symlink.exists()` returns `False`, but `broken_symlink.symlink_to()` will fail).

I tried to install a different workflow with the same name, and it raises
```
$ cylc install -C cylc-forecasting-workflow/
WorkflowFilesError: Source directory between runs are not consistent.
```
so maybe a similar nicely-displayed error should be raised

Comment (datamel) [reply cylc/flow/workflow_files.py:1356]: I'm struggling to reproduce this case, I suspect I have misunderstood. 
I have installed a workflow from `~/cylc-src/test-workflow` into `~/cylc-run/test-workflow/run1`
I then move `~/cylc-src/test-workflow` to `~/cylc-src/test-workflow2`. The symlink, as expected, breaks.
Then :
```
$pwd
~/cylc-src
$ cylc install -C test-workflow
WorkflowFilesError: no flow.cylc or suite.rc in test-workflow
```

I have also tried reinstalling:

```
$ cylc reinstall test-workflow/run1
WorkflowFilesError: Workflow source dir is not accessible: "~/cylc-src/test-workflow".
Restore the source or modify the ~/cylc-run/test-workflow/_cylc-install/source" symlink to continue. 
```
If I install the moved `test-workflow2`, it installs as a new flow, as expected, into  `~/cylc-run/test-workflow2/run1`

Comment (MetRonnie) [reply cylc/flow/workflow_files.py:1356]: Try:
- Create `~/cylc-src/test/flow.cylc`
- `$ cylc install test`
- Delete the `~/cylc-src/test/` dir
- Create `~/Documents/test/flow.cylc` (or wherever, doesn't have to be in Documents).
- `$ cylc install -C ~/Documents/test`

Comment (datamel) [reply cylc/flow/workflow_files.py:1363]: Got it, thanks for clarifying. Should be fixed now with a test.

File: cylc/flow/workflow_files.py [hunk 2]
```diff
@@ -1352,8 +1352,15 @@ def install_workflow(
     # check source link matches the source symlink from workflow dir.
     cylc_install.mkdir(parents=True, exist_ok=True)
     if not source_link.exists():
+        if source_link.is_symlink():
+            raise WorkflowFilesError(
+                'Workflow run directory contains a broken source symlink: '
+                f'{source_link}\n'
+                f'Restore the original source or modify this'
+                ' symlink, then try cylc install again.'
+            )
```
Comment (MetRonnie) [inline cylc/flow/workflow_files.py:1361]: I feel like the fact that the symlink is broken shouldn't be the main point, it's that `Workflow source dir is not accessible` (from the other error message). In fact, I think the other error message can be reused for this?

It makes me wonder if we should have a cli option for `cylc install` to "re-source" the workflow from the given directory (which would be a follow-up issue).

**Comment by MetRonnie** [comment]
Just ran into this bug again, would be good to get it resolved soon

**Review by hjoliver — APPROVED** [review]
Looks good, tested as working :+1:

**Review by MetRonnie — APPROVED** [review]
Happy with it to merge.

However, going back to the example I commented earlier (sorry!)
>- Create `~/cylc-src/test/flow.cylc`
>- `$ cylc install test`
>- Delete the `~/cylc-src/test/` dir
>- Create `~/Documents/test/flow.cylc` (or wherever, doesn't have to be in Documents).
>- `$ cylc install -C ~/Documents/test`

The warning message
```
WorkflowFilesError: Workflow source dir is not accessible: ~/cylc-src/test".
Restore the source or modify the "~/cylc-run/test/_cylc-install/source"
symlink to continue.
```

might not be so helpful in this case, and I wonder if the `Source directory not consistent between runs` message should take precedence instead.

Up to you if you want to address this

**Comment by hjoliver** [comment]
I'm gonna merge this - thanks @datamel :+1:  and post a "small" issue to review the wording of installation-related error messages.

**Pull request merged**